File: src/jarvis/agent/retry.py
"""Retry/replan helpers for AgentLoop."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from .types import ToolCall, ToolResult


@dataclass
class ErrorClassification:
    category: str
    retryable: bool
    replan: bool
    reason: str


class ErrorClassifier:
    def classify(self, tool_result: ToolResult) -> ErrorClassification:
        err = str(tool_result.error or "").lower()
        if not err:
            return ErrorClassification(category="none", retryable=False, replan=False, reason="no_error")
        if "timeout" in err:
            return ErrorClassification(category="timeout", retryable=True, replan=False, reason="command_timeout")
        if "approval_required" in err or "denied" in err or "permission" in err:
            return ErrorClassification(category="permission", retryable=False, replan=True, reason="permission_denied")
        if "unknown tool" in err or "unknown_tool" in err:
            return ErrorClassification(category="tool_schema", retryable=True, replan=True, reason="unknown_tool")
        if "not found" in err or "does_not_exist" in err or "no such file" in err:
            return ErrorClassification(category="not_found", retryable=False, replan=True, reason="missing_target")
        if "invalid" in err or "malformed" in err or "parameter" in err:
            return ErrorClassification(category="bad_params", retryable=True, replan=True, reason="invalid_parameters")
        if "assertion" in err or "test" in err:
            return ErrorClassification(category="test_failed", retryable=False, replan=True, reason="tests_failed")
        return ErrorClassification(category="other", retryable=False, replan=True, reason="tool_failed")


class RetryPolicy:
    def __init__(self, *, max_retries: int = 2) -> None:
        self.max_retries = max_retries
        self.retry_counts: dict[str, int] = {}

    def should_retry(self, call: ToolCall, classification: ErrorClassification) -> bool:
        if not classification.retryable:
            return False
        key = f"{call.name}:{classification.category}"
        used = self.retry_counts.get(key, 0)
        if used >= self.max_retries:
            return False
        self.retry_counts[key] = used + 1
        return True


@dataclass
class FailureRecord:
    tool_name: str
    error_category: str
    error_message: str
    step: int


class FailureTracker:
    """Track failures and repetition to detect infinite loops."""

    def __init__(self, *, max_consecutive: int = 5, max_same_tool: int = 4, max_repeat: int = 3) -> None:
        self.max_consecutive = max_consecutive
        self.max_same_tool = max_same_tool
        self.max_repeat = max_repeat
        self.consecutive_failures: list[FailureRecord] = []
        self.tool_failure_counts: dict[str, int] = {}
        self.tool_total_calls: dict[str, int] = {}
        self._synthesis_nudged: set[str] = set()

    def record_failure(self, tool_name: str, error_category: str, error_message: str, step: int) -> None:
        self.consecutive_failures.append(
            FailureRecord(tool_name=tool_name, error_category=error_category,
                          error_message=error_message, step=step)
        )
        self.tool_failure_counts[tool_name] = self.tool_failure_counts.get(tool_name, 0) + 1
        self.tool_total_calls[tool_name] = self.tool_total_calls.get(tool_name, 0) + 1

    def record_success(self, tool_name: str) -> None:
        self.consecutive_failures.clear()
        self.tool_failure_counts.pop(tool_name, None)
        self.tool_total_calls[tool_name] = self.tool_total_calls.get(tool_name, 0) + 1

    def should_stop(self) -> tuple[bool, str]:
        """Check if we should stop due to consecutive failures. Returns (stop, reason)."""
        if len(self.consecutive_failures) >= self.max_consecutive:
            last = self.consecutive_failures[-1]
            return True, (
                f"{self.max_consecutive} consecutive tool failures. "
                f"Last error ({last.tool_name}): {last.error_message[:200]}"
            )
        return False, ""

    def should_reject_tool(self, tool_name: str) -> tuple[bool, str, str]:
        """Returns (reject, reason, kind) where kind is 'failure' (too many errors)
        or 'repeat' (too many total calls — model should synthesize instead)."""
        fail_count = self.tool_failure_counts.get(tool_name, 0)
        if fail_count >= self.max_same_tool:
            return True, (
                f"Tool `{tool_name}` has failed {fail_count} times; try a different approach or stop."
            ), "failure"
        total = self.tool_total_calls.get(tool_name, 0)
        if total >= self.max_repeat:
            return True, (
                f"Tool `{tool_name}` has been called {total} times this turn. "
                "Stop calling it and synthesize a final answer from the results you already have."
            ), "repeat"
        return False, "", ""

    def is_repeat_hard_stop(self, tool_name: str) -> bool:
        """Returns True if this tool already received a synthesis nudge and
        the model is still trying to call it — hard stop is warranted."""
        if tool_name in self._synthesis_nudged:
            return True
        self._synthesis_nudged.add(tool_name)
        return False


class ReplanPolicy:
    def __init__(self, *, max_replans: int = 2) -> None:
        self.max_replans = max_replans
        self.replan_count = 0

    def should_replan(self, classification: ErrorClassification) -> bool:
        if not classification.replan:
            return False
        if self.replan_count >= self.max_replans:
            return False
        self.replan_count += 1
        return True

    def build_replan_observation(self, tool_result: ToolResult, classification: ErrorClassification) -> dict[str, Any]:
        return {
            "event": "replan_hint",
            "tool_name": tool_result.name,
            "category": classification.category,
            "reason": classification.reason,
            "error": tool_result.error,
        }

File: packages/agent/src/retry.ts
// ============================================================================
// Retry utilities — jittered backoff, error classification, failure tracking
// ============================================================================

import type { ToolCall, ToolResult } from '@jarvis/shared';

// ============================================================================
// Jittered Backoff
// ============================================================================

/**
 * Calculate delay with jittered exponential backoff.
 *
 * Formula: min(base * 2^(attempt-1), maxDelay) * (1 - jitterRatio * random)
 *
 * @param attempt  1-indexed attempt number (1 = first retry after initial failure)
 * @param baseDelay  Base delay in ms, default 5000
 * @param maxDelay   Maximum delay in ms, default 120000
 * @param jitterRatio  Fraction of delay to randomize, default 0.3 (30%)
 */
export function jitteredBackoff(
  attempt: number,
  baseDelay = 5_000,
  maxDelay = 120_000,
  jitterRatio = 0.3,
): number {
  const exponential = baseDelay * Math.pow(2, attempt - 1);
  const capped = Math.min(exponential, maxDelay);
  const jitter = capped * jitterRatio * Math.random();
  return capped - jitter;
}

// ============================================================================
// Retry Configuration
// ============================================================================

export interface RetryConfig {
  /** Maximum number of retry attempts (not including the initial call) */
  maxRetries: number;
  /** Base delay in ms between retries, default 5000 */
  baseDelay?: number;
  /** Maximum delay in ms between retries, default 120000 */
  maxDelay?: number;
  /** HTTP status codes that trigger a retry */
  retryOn?: number[];
}

// ============================================================================
// withRetry
// ============================================================================

/**
 * Execute an async function with retry logic.
 *
 * Retries on:
 * - Any error matching shouldRetry (if provided)
 * - HTTP errors with status codes in config.retryOn
 * - Network/timeout errors (by default)
 *
 * Uses jittered exponential backoff between attempts.
 */
export async function withRetry<T>(
  fn: () => Promise<T>,
  config: RetryConfig,
  shouldRetry?: (error: unknown) => boolean,
): Promise<T> {
  const baseDelay = config.baseDelay ?? 5_000;
  const maxDelay = config.maxDelay ?? 120_000;
  const retryOn = config.retryOn ?? [429, 500, 502, 503, 504];

  let lastError: unknown;

  for (let attempt = 0; attempt <= config.maxRetries; attempt++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error;

      // If this was the last attempt, rethrow
      if (attempt >= config.maxRetries) {
        throw error;
      }

      // Check if we should retry this error
      const shouldRetryThis =
        shouldRetry?.(error) ??
        _isRetryableError(error, retryOn) ??
        true;

      if (!shouldRetryThis) {
        throw error;
      }

      // Wait before retrying
      const delay = jitteredBackoff(attempt + 1, baseDelay, maxDelay);
      await _sleep(delay);
    }
  }

  // Should never reach here, but satisfy TypeScript
  throw lastError;
}

// ============================================================================
// Internal helpers
// ============================================================================

function _sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Determine if an error is retryable based on common error shapes.
 */
function _isRetryableError(error: unknown, retryOn: number[]): boolean | null {
  if (error && typeof error === 'object') {
    const err = error as Record<string, unknown>;

    // Check HTTP status code on error object (OpenAI SDK pattern)
    if (typeof err['status'] === 'number') {
      if (retryOn.includes(err['status'] as number)) return true;
      return false;
    }

    if (err['error'] && typeof err['error'] === 'object') {
      const inner = err['error'] as Record<string, unknown>;
      if (typeof inner['status'] === 'number') {
        if (retryOn.includes(inner['status'] as number)) return true;
        return false;
      }
    }

    if (typeof err['code'] === 'string') {
      const code = err['code'] as string;
      if (['ECONNRESET', 'ETIMEDOUT', 'ECONNREFUSED', 'ENOTFOUND'].includes(code)) {
        return true;
      }
    }
  }

  if (error instanceof TypeError && error.message.includes('fetch failed')) {
    return true;
  }

  // Return null to defer to the default (retry unknown errors)
  return null;
}

// ============================================================================
// Error Classification
// ============================================================================

export interface ErrorClassification {
  category: string;
  retryable: boolean;
  replan: boolean;
  reason: string;
}

export class ErrorClassifier {
  classify(toolResult: ToolResult): ErrorClassification {
    const err = (toolResult.error ?? '').toLowerCase();
    if (!err) {
      return { category: 'none', retryable: false, replan: false, reason: 'no_error' };
    }
    if (err.includes('timeout')) {
      return { category: 'timeout', retryable: true, replan: false, reason: 'command_timeout' };
    }
    if (err.includes('approval_required') || err.includes('denied') || err.includes('permission')) {
      return { category: 'permission', retryable: false, replan: true, reason: 'permission_denied' };
    }
    if (err.includes('unknown tool') || err.includes('unknown_tool')) {
      return { category: 'tool_schema', retryable: true, replan: true, reason: 'unknown_tool' };
    }
    if (err.includes('not found') || err.includes('does_not_exist') || err.includes('no such file')) {
      return { category: 'not_found', retryable: false, replan: true, reason: 'missing_target' };
    }
    if (err.includes('invalid') || err.includes('malformed') || err.includes('parameter')) {
      return { category: 'bad_params', retryable: true, replan: true, reason: 'invalid_parameters' };
    }
    if (err.includes('assertion') || err.includes('test')) {
      return { category: 'test_failed', retryable: false, replan: true, reason: 'tests_failed' };
    }
    return { category: 'other', retryable: false, replan: true, reason: 'tool_failed' };
  }
}

// ============================================================================
// RetryPolicy
// ============================================================================

export class RetryPolicy {
  private maxRetries: number;
  private retryCounts = new Map<string, number>();

  constructor(maxRetries = 2) {
    this.maxRetries = maxRetries;
  }

  shouldRetry(call: ToolCall, classification: ErrorClassification): boolean {
    if (!classification.retryable) return false;
    const key = `${call.name}:${classification.category}`;
    const used = this.retryCounts.get(key) ?? 0;
    if (used >= this.maxRetries) return false;
    this.retryCounts.set(key, used + 1);
    return true;
  }
}

// ============================================================================
// FailureTracker
// ============================================================================

export interface FailureRecord {
  toolName: string;
  errorCategory: string;
  errorMessage: string;
  step: number;
}

export class FailureTracker {
  private maxConsecutive: number;
  private maxSameTool: number;
  private maxRepeat: number;
  consecutiveFailures: FailureRecord[] = [];
  toolFailureCounts = new Map<string, number>();
  toolTotalCalls = new Map<string, number>();
  private _synthesisNudged = new Set<string>();

  constructor(maxConsecutive = 5, maxSameTool = 4, maxRepeat = 3) {
    this.maxConsecutive = maxConsecutive;
    this.maxSameTool = maxSameTool;
    this.maxRepeat = maxRepeat;
  }

  recordFailure(
    toolName: string,
    errorCategory: string,
    errorMessage: string,
    step: number,
  ): void {
    this.consecutiveFailures.push({ toolName, errorCategory, errorMessage, step });
    this.toolFailureCounts.set(toolName, (this.toolFailureCounts.get(toolName) ?? 0) + 1);
    this.toolTotalCalls.set(toolName, (this.toolTotalCalls.get(toolName) ?? 0) + 1);
  }

  recordSuccess(toolName: string): void {
    this.consecutiveFailures = [];
    this.toolFailureCounts.delete(toolName);
    this.toolTotalCalls.set(toolName, (this.toolTotalCalls.get(toolName) ?? 0) + 1);
  }

  shouldStop(): { stop: boolean; reason: string } {
    if (this.consecutiveFailures.length >= this.maxConsecutive) {
      const last = this.consecutiveFailures[this.consecutiveFailures.length - 1];
      return {
        stop: true,
        reason: `${this.maxConsecutive} consecutive tool failures. Last error (${last.toolName}): ${last.errorMessage.slice(0, 200)}`,
      };
    }
    return { stop: false, reason: '' };
  }

  shouldRejectTool(toolName: string): { reject: boolean; reason: string; kind: string } {
    const failCount = this.toolFailureCounts.get(toolName) ?? 0;
    if (failCount >= this.maxSameTool) {
      return {
        reject: true,
        reason: `Tool \`${toolName}\` has failed ${failCount} times; try a different approach or stop.`,
        kind: 'failure',
      };
    }
    const total = this.toolTotalCalls.get(toolName) ?? 0;
    if (total >= this.maxRepeat) {
      return {
        reject: true,
        reason: `Tool \`${toolName}\` has been called ${total} times this turn. Stop calling it and synthesize a final answer from the results you already have.`,
        kind: 'repeat',
      };
    }
    return { reject: false, reason: '', kind: '' };
  }

  isRepeatHardStop(toolName: string): boolean {
    if (this._synthesisNudged.has(toolName)) return true;
    this._synthesisNudged.add(toolName);
    return false;
  }
}

// ============================================================================
// ReplanPolicy
// ============================================================================

export class ReplanPolicy {
  private maxReplans: number;
  replanCount = 0;

  constructor(maxReplans = 2) {
    this.maxReplans = maxReplans;
  }

  shouldReplan(classification: ErrorClassification): boolean {
    if (!classification.replan) return false;
    if (this.replanCount >= this.maxReplans) return false;
    this.replanCount++;
    return true;
  }

  buildReplanObservation(
    toolResult: ToolResult,
    classification: ErrorClassification,
  ): Record<string, unknown> {
    return {
      event: 'replan_hint',
      tool_name: toolResult.name,
      category: classification.category,
      reason: classification.reason,
      error: toolResult.error,
    };
  }
}
